### PR TITLE
Arithmetic

### PR DESCRIPTION
**Pull request by codygustafson**

### CODE REVIEW

**Comment by wlmcewen** [comment]
There's a lot to unpack here. Let's get a description on what we're trying to accomplish in this PR.

File: lib/sparkql/parser.y [hunk 1]
```diff
@@ -28,6 +28,8 @@ class Sparkql::Parser
 # Unary minus is always tied to value, such as for negative numbers.
 prechigh
   nonassoc UMINUS
+  left MUL DIV MOD
+  left ADD SUB
```
Comment (wlmcewen) [inline lib/sparkql/parser.y]: @eeggers: here's how you handle precedence outside of grammar terminals.

File: lib/sparkql/parser.y [hunk 2]
```diff
@@ -88,11 +90,18 @@ rule
 # Keyword for searching on, these fields should be discovered using the metadata 
 # rules. In general, Keywords that cannot be found will be dropped from the 
 # filter.
+
   field
-  	: STANDARD_FIELD
-  	| CUSTOM_FIELD
-  	| function
-  	;
+    : field ADD field { result = tokenize_arithmetic(val[0], val[1], val[2]) }
```
Comment (wlmcewen) [inline lib/sparkql/parser.y]: This doesn't sit right with me. I'd extract an `arithmetic_expression` terminal out of this or incorporate this with "field" some other way. I guess I started this with `function` and I apologize.

In any case, a bunch of these things are no longer fields, and we should update / doc the terminal accordingly.

Comment (codygustafson) [reply lib/sparkql/parser.y]: 👍 Yeah, that sounds good to me.

File: GRAMMAR.md [hunk 3]
```diff
@@ -92,10 +94,16 @@ filter.
...
+       : field ADD field 
+       | field SUB field 
+       | field MUL field 
+       | field DIV field 
+       | field MOD field 
+       | STANDARD_FIELD
+       | CUSTOM_FIELD
+       | function
+       | numeric
+       ;
```
Comment (codygustafson) [inline GRAMMAR.md]: A tricky part with these is that the field side  and the condition side are becoming more and more tide together. Previously the only place literals were allowed was within the functions, now we will be requiring it within arithmetic as well.

**Comment by codygustafson** [comment]
@wlmcewen, could you look this one over for me again?

File: lib/sparkql/parser.y [hunk 4]
```diff
@@ -52,77 +52,99 @@ rule
...
-    | field SUB field { result = tokenize_arithmetic(val[0], val[1], val[2]) }
-    | field MUL field { result = tokenize_arithmetic(val[0], val[1], val[2]) }
-    | field DIV field { result = tokenize_arithmetic(val[0], val[1], val[2]) }
-    | field MOD field { result = tokenize_arithmetic(val[0], val[1], val[2]) }
-    | STANDARD_FIELD
+    : STANDARD_FIELD
     | CUSTOM_FIELD
+    ;
+
+  arithmetic_expression
```
Comment (wlmcewen) [inline lib/sparkql/parser.y]: I'd rename this to have the word `field` in it. The point being is we can have arithmetic with the field or arithmetic with the condition, and never the twain shall meet.

File: lib/sparkql/parser.y [hunk 5]
```diff
@@ -52,77 +52,99 @@ rule
...
+    : field_function_expression arithmetic literal { result = tokenize_arithmetic(val[0], val[1], val[2]) }
+    | literal arithmetic field_function_expression { result = tokenize_arithmetic(val[0], val[1], val[2]) }
+    | field_function_expression arithmetic literals { result = tokenize_arithmetic(val[0], val[1], val[2]) }
+    | literals arithmetic field_function_expression { result = tokenize_arithmetic(val[0], val[1], val[2]) }
+    ;
+
+  arithmetic
```
Comment (wlmcewen) [inline lib/sparkql/parser.y]: I'd kick this part down towards the bottom with the `literals`

File: lib/sparkql/parser.y (hunk 4 above)
Comment (codygustafson) [reply lib/sparkql/parser.y]: Good call 👍

**Comment by codygustafson** [comment]
Where do you want that documented? In the grammar file itself?

**Comment by wlmcewen** [comment]
> Where do you want that documented? In the grammar file itself?

Exactly. In case you haven't noticed, the `parser.y` file generates the `GRAMMAR.md` file, which is pulled into http://sparkplatform.com/docs/supporting_documentation/sparkql_grammar automatically.

**Comment by codygustafson** [comment]
Yeah, just wanted to be sure there wasn't another location I should be looking 👍

**Pull request merged**